cpp/src/barretenberg/honk/pcs/kzg/kzg.hpp

```diff
@@ -1,93 +1,65 @@
-
-  public:
-    /**
-     * @brief Construct a new Bilinear Accumulator object given a claim (C,r,v) and proof π.
-     *      - P₀ = C − v⋅[1]₁ + r⋅[x]₁
-     *      - P₁ = −π
-     * @param claim an OpeningClaim (C,r,v)
-     * @param proof a Commitment π
-     */
-    BilinearAccumulator(const OpeningClaim<Params>& claim, const Commitment& proof)
-        : lhs(claim.commitment - (Commitment::one() * claim.opening_pair.evaluation) +
-              (proof * claim.opening_pair.challenge))
-        , rhs(-proof)
-    {}
-
-    /**
-     * @brief verifies the accumulator with a pairing check
-     *
-     * @param vk VerificationKey
-     * @return e(P₀,[1]₁)e(P₁,[x]₂)≡ [1]ₜ
-     */
-    bool verify(std::shared_ptr<VK> vk) const { return vk->pairing_check(lhs, rhs); };
-
-    bool operator==(const BilinearAccumulator& other) const = default;
-
-    CommitmentAffine lhs, rhs;
-};
-
-template <typename Params> class UnivariateOpeningScheme {
-    using CK = typename Params::CK;
 
+template <typename Params> class KZG {
+    using CK = typename Params::CommitmentKey;
+    using VK = typename Params::VerificationKey;
     using Fr = typename Params::Fr;
     using Commitment = typename Params::Commitment;
-    using CommitmentAffine = typename Params::C;
+    using GroupElement = typename Params::GroupElement;
     using Polynomial = barretenberg::Polynomial<Fr>;
 
-  public:
-    using Accumulator = BilinearAccumulator<Params>;
-
     /**
-     * @brief Compute KZG opening proof polynomial
+     * @brief Computes the KZG commitment to an opening proof polynomial at a single evaluation point
      *
-     * @param opening_pair OpeningPair = {r, v = polynomial(r)}
-     * @param polynomial the witness polynomial being opened
-     * @return KZG quotient polynomial of the form (p(X) - v) / (X - r)
+     * @param ck The commitment key which has a commit function, the srs and pippenger_runtime_state
+     * @param opening_pair OpeningPair = {r, v = p(r)}
+     * @param polynomial The witness whose opening proof needs to be computed
+     * @param prover_transcript Prover transcript
      */
-    static Polynomial compute_opening_proof_polynomial(const OpeningPair<Params>& opening_pair,
-                                                       const Polynomial& polynomial)
+  public:
+    static void compute_opening_proof(std::shared_ptr<CK> ck,
+                                      const OpeningPair<Params>& opening_pair,
+                                      const Polynomial& polynomial,
+                                      ProverTranscript<Fr>& prover_trancript)
     {
         Polynomial quotient(polynomial);
         quotient[0] -= opening_pair.evaluation;
+        // Computes the coefficients for the quotient polynomial q(X) = (p(X) - v) / (X - r) through an FFT
         quotient.factor_roots(opening_pair.challenge);
-
-        return quotient;
+        auto quotient_commitment = ck->commit(quotient);
```

Comment:
Just flagging this: Historically we have had to rely on the work_queue to compute commitments - we could not compute them directly inside of other functions. (This is due to the way we manually achieve parallelization in WASM). However with Charlie's recent "low memory prover" work, he's updated to WASI 20 (?) which has some native support for multithreading, eliminating the need for the work_queue. My understanding is that his work depends on some fairly experimental libraries so I think it's too early to say for sure whether we can say goodbye to the work_queue forever. All of this is to say that I think it's ok to compute this commitment directly for now but it's probably worth leaving a note just in case this has to be adjusted in the future.

Reply:
That’s a very good point, thank you. Thought I left a comment